File: web/src/gnn.js
// Runtime for the graph model, in plain JavaScript.
//
// The network is 42k parameters of small MLPs plus gather/scatter over ~42
// directed edges, so it needs no ONNX runtime, no WASM, and no network fetch
// beyond the weights themselves. Measured ~2.3ms per message-passing round,
// so ~23ms for a full-strength move -- invisible when the game asks for one
// prediction per round, but do not put this in a per-frame loop.
//
// This file must compute exactly what trussnet/gnn.py computes. That is not a
// hope -- training/verify_export.py checks the two against each other on real
// trusses and fails loudly if they drift.

function decode(b64) {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  // Little-endian float32, matching the '<f4' the exporter writes. Every
  // platform that runs a browser is little-endian, so the view is free.
  return new Float32Array(bytes.buffer);
}

function decodeMLP(layers) {
  return layers.map((layer) =>
    layer.type === 'linear'
      ? { type: 'linear', in: layer.in, out: layer.out, w: decode(layer.weight), b: decode(layer.bias) }
      : { type: layer.type }
  );
}

// y = x @ W^T + b, for `rows` stacked inputs. W is row-major (out, in), which
// is how torch stores nn.Linear.
function linear(layer, x, rows) {
  const { in: nin, out: nout, w, b } = layer;
  const y = new Float32Array(rows * nout);
  for (let r = 0; r < rows; r++) {
    const xo = r * nin;
    const yo = r * nout;
    for (let o = 0; o < nout; o++) {
      let acc = b[o];
      const wo = o * nin;
      for (let i = 0; i < nin; i++) acc += w[wo + i] * x[xo + i];
      y[yo + o] = acc;
    }
  }
  return y;
}

function silu(x) {
  for (let i = 0; i < x.length; i++) x[i] = x[i] / (1 + Math.exp(-x[i]));
  return x;
}

function runMLP(layers, x, rows) {
  let out = x;
  for (const layer of layers) {
    out = layer.type === 'linear' ? linear(layer, out, rows) : silu(out);
  }
  return out;
}

// Normalises each row over its features, matching torch.nn.LayerNorm -- which
// uses the biased variance, not the sample variance.
function layerNorm(x, rows, dim, weight, bias, eps) {
  for (let r = 0; r < rows; r++) {
    const o = r * dim;
    let mean = 0;
    for (let i = 0; i < dim; i++) mean += x[o + i];
    mean /= dim;
    let variance = 0;
    for (let i = 0; i < dim; i++) {
      const d = x[o + i] - mean;
      variance += d * d;
    }
    variance /= dim;
    const inv = 1 / Math.sqrt(variance + eps);
    for (let i = 0; i < dim; i++) x[o + i] = (x[o + i] - mean) * inv * weight[i] + bias[i];
  }
  return x;
}

export class TrussGNN {
  constructor(payload) {
    if (payload.format !== 'trussgnn/1') {
      throw new Error(`unsupported model format ${payload.format}`);
    }
    this.hidden = payload.hidden;
    this.rounds = payload.rounds;
    this.sigma = payload.sigma;
    this.valMean = payload.valMean;
    this.eps = payload.layerNormEps;
    this.nodeEncoder = decodeMLP(payload.nodeEncoder);
    this.edgeEncoder = decodeMLP(payload.edgeEncoder);
    this.message = decodeMLP(payload.message);
    this.update = decodeMLP(payload.update);
    this.decoder = decodeMLP(payload.decoder);
    this.normWeight = decode(payload.normWeight);
    this.normBias = decode(payload.normBias);
  }

  static async load(url) {
    const response = await fetch(url);
    if (!response.ok) throw new Error(`could not load model from ${url}: ${response.status}`);
    return new TrussGNN(await response.json());
  }

  /**
   * @param graph  {nodeFeat, edgeIndex, edgeFeat, freeMask, nodes, edges}
   * @param rounds how many message-passing rounds to run. Fewer than the model
   *   was trained with is a less-converged solver, which is the difficulty
   *   dial: 1 round plays at ~25%, 6 at ~68%, 10 at ~96%.
   * @returns Float32Array(nodes * 2) of displacements, in units of the support
   *   span the caller measured.
   */
  predict(graph, rounds = this.rounds) {
    const { nodeFeat, edgeIndex, edgeFeat, freeMask, nodes, edges } = graph;
    const H = this.hidden;

    let h = runMLP(this.nodeEncoder, Float32Array.from(nodeFeat), nodes);
    const e = runMLP(this.edgeEncoder, Float32Array.from(edgeFeat), edges);

    const catIn = new Float32Array(edges * 3 * H);
    const updIn = new Float32Array(nodes * 2 * H);

    for (let t = 0; t < rounds; t++) {
      for (let k = 0; k < edges; k++) {
        const src = edgeIndex[k * 2] * H;
        const dst = edgeIndex[k * 2 + 1] * H;
        const o = k * 3 * H;
        for (let i = 0; i < H; i++) {
          catIn[o + i] = h[src + i];
          catIn[o + H + i] = h[dst + i];
          catIn[o + 2 * H + i] = e[k * H + i];
        }
      }
      const m = runMLP(this.message, catIn, edges);

      // Sum aggregation at the destination node: assembling K sums every
      // member meeting at a node, so a busier node really is stiffer.
      const agg = new Float32Array(nodes * H);
      for (let k = 0; k < edges; k++) {
        const dst = edgeIndex[k * 2 + 1] * H;
        const o = k * H;
        for (let i = 0; i < H; i++) agg[dst + i] += m[o + i];
      }

      for (let n = 0; n < nodes; n++) {
        for (let i = 0; i < H; i++) {
          updIn[n * 2 * H + i] = h[n * H + i];
          updIn[n * 2 * H + H + i] = agg[n * H + i];
        }
      }
      const delta = runMLP(this.update, Float32Array.from(updIn), nodes);
      for (let i = 0; i < h.length; i++) h[i] += delta[i];
      h = layerNorm(h, nodes, H, this.normWeight, this.normBias, this.eps);
    }

    const out = runMLP(this.decoder, h, nodes);
    // Supports are pinned; the boundary condition is imposed, never predicted.
    for (let n = 0; n < nodes; n++) {
      out[n * 2] *= freeMask[n];
      out[n * 2 + 1] *= freeMask[n];
    }
    return out;
  }
}

File: training/trussnet/gnn.py
"""A graph network that learns the truss solve, shaped like the solver itself.

Why message passing rather than convolution
-------------------------------------------
The answer is u = K^-1 f. K is sparse -- one block per member -- but its inverse
is dense: every node's displacement depends on every other node. A feedforward
local operator is the wrong computational class for that, which is why widening
the CNN's receptive field to cover the support span made it *worse* rather than
better (measured 2026-08-18: 405px field scored 0.67 points below a 237px one).

Iterative solvers for K u = f -- Jacobi, Gauss-Seidel, conjugate gradient -- are
message passing on K's sparsity graph. So the processor here is that, learned:
T rounds of exchange along the members, with weights *shared* across rounds so
a round is genuinely an iteration rather than a layer. Two consequences follow.
Running fewer rounds at inference than in training yields a less-converged, and
therefore weaker but still physically coherent, opponent -- difficulty as solver
iterations. And the graph diameter is 3, so information crosses any truss in 3
rounds; more rounds buy accuracy, not reach.

The edge features are the exact entries of the element stiffness matrix,
(1/L_hat) * [[c^2, cs], [cs, s^2]]. The network is handed the ingredients of K
and has only to learn to invert it, never to rediscover it from geometry.
"""
from __future__ import annotations

import torch
import torch.nn as nn

from .graph_data import EDGE_FEATURES, NODE_FEATURES


def mlp(sizes, out_activation=False):
    layers = []
    for i in range(len(sizes) - 1):
        layers.append(nn.Linear(sizes[i], sizes[i + 1]))
        if i < len(sizes) - 2 or out_activation:
            layers.append(nn.SiLU())
    return nn.Sequential(*layers)


class TrussGNN(nn.Module):
    def __init__(self, hidden=64, rounds=10, shared=True):
        super().__init__()
        self.rounds, self.shared = rounds, shared

        self.node_encoder = mlp([NODE_FEATURES, hidden, hidden], out_activation=True)
        self.edge_encoder = mlp([EDGE_FEATURES, hidden, hidden], out_activation=True)

        blocks = 1 if shared else rounds
        self.message = nn.ModuleList(
            mlp([3 * hidden, hidden, hidden], out_activation=True) for _ in range(blocks)
        )
        self.update = nn.ModuleList(
            mlp([2 * hidden, hidden, hidden], out_activation=True) for _ in range(blocks)
        )
        # LayerNorm keeps an unrolled, weight-shared recurrence from drifting in
        # scale over its iterations -- the failure mode a deep shared stack has
        # that a plain feedforward net does not.
        self.norm = nn.LayerNorm(hidden)
        self.decoder = mlp([hidden, hidden, 2])

    def forward(self, batch, rounds=None):
        rounds = self.rounds if rounds is None else rounds
        node_feat = batch['node_feat']
        edge_index, edge_feat = batch['edge_index'], batch['edge_feat']
        edge_mask = batch['edge_mask'].unsqueeze(-1)

        h = self.node_encoder(node_feat)
        e = self.edge_encoder(edge_feat) * edge_mask

        src = edge_index[..., 0].unsqueeze(-1).expand(-1, -1, h.shape[-1])
        dst = edge_index[..., 1].unsqueeze(-1).expand(-1, -1, h.shape[-1])

        for t in range(rounds):
            block = 0 if self.shared else t
            m = self.message[block](torch.cat([h.gather(1, src), h.gather(1, dst), e], dim=-1))
            # Sum, not mean: assembling K sums the contribution of every member
            # meeting at a node, so a node with more members really is stiffer.
            agg = torch.zeros_like(h).scatter_add_(1, dst, m * edge_mask)
            h = self.norm(h + self.update[block](torch.cat([h, agg], dim=-1)))

        # Supports are fully pinned. That is a boundary condition, not something
        # to be learned approximately, so it is imposed exactly.
        return self.decoder(h) * batch['free_mask'].unsqueeze(-1)


def physics_residual(u_hat, batch):
    """How far a predicted displacement field is from equilibrium.

    Needs no labels: it is the non-dimensionalised K u - f evaluated on the
    graph the model was given. Returns the per-node residual for free nodes,
    scaled by the applied load so it reads as a fraction of that load.
    """
    edge_index, edge_feat = batch['edge_index'], batch['edge_feat']
    edge_mask = batch['edge_mask'].unsqueeze(-1)
    src, dst = edge_index[..., 0], edge_index[..., 1]

    idx_src = src.unsqueeze(-1).expand(-1, -1, 2)
    idx_dst = dst.unsqueeze(-1).expand(-1, -1, 2)
    rel = u_hat.gather(1, idx_dst) - u_hat.gather(1, idx_src)

    cc, cs, ss, inv_len = (edge_feat[..., i] for i in range(4))
    # (1 / L_hat) * [[c^2, cs], [cs, s^2]] @ (u_j - u_i), the member's pull on
    # the node it points at.
    fx = inv_len * (cc * rel[..., 0] + cs * rel[..., 1])
    fy = inv_len * (cs * rel[..., 0] + ss * rel[..., 1])
    internal = torch.stack([fx, fy], dim=-1) * edge_mask

    # `internal` accumulated at dst is (K u)_i; equilibrium wants it to equal
    # the applied load, so the residual subtracts. Verified against the exact
    # field: this form gives ~1e-6, the opposite sign gives exactly 2f.
    net = torch.zeros_like(u_hat).scatter_add_(1, idx_dst, internal)
    external = batch['node_feat'][..., 3:5]
    residual = (net - external) * batch['free_mask'].unsqueeze(-1)
    return residual / batch['f_hat'].view(-1, 1, 1)


def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)
